### PR TITLE
wip: chore: minor improvements; mainly: dropped g_syscall_code_routing_table and added an m4 script to automatically create syscall tables (normal + ia32)

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind cleanup

**Any specific area of the project related to this PR?**

/area libscap

**What this PR does / why we need it**:

* added a flexible array member to struct ppm_evt_hdr to manage events payload
Std states:
> As a special case, the last element of a structure with more than one named member may have an incomplete array type; this is called a flexible array member. In most situations, the flexible array member is ignored. In particular, the size of the structure is as if the flexible array member were omitted except that it may have more trailing padding than the omission would imply.

Note about the padding: we use an uint8_t for flexible array member, thus it shall not change padding in our case: you store sizeof(ppm_evt_hdr) + sizeof hdrs + sizeof(payload) and you load exactly the same;
but the flexible array member helps in avoiding pointer arithmetic when dealing with hdrs and payload.

I tested the dump/read of scap files:
1. new version from new version
2. new version from old version
3. old version from new version  

* merged `g_syscall_code_routing_table` into `g_syscall_table`

The `g_syscall_code_routing_table` was not really useful by itself, and the 2 tables had same indexing.

* dropped syscall_table.c in favour of syscall_table.m4 script that will automatically build syscall_table.c
at compile time. It allows adding a new syscall by just inserting a one-liner in the new `X_SYSCALLS` X macro.

It uses an m4 script; i think that m4 is the most widely available solution.  
Why does this latter change matter? Because we were already hit by issues like forgetting to populate ia32 syscalls, or forgetting to populate `PPM_SC_` code (the latter case is already somewhat mitigated by above point).

* forced gnu99 as C standard

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by FedeDP** [comment]
Rebased on master.

**Comment by leogr** [comment]
Closing and reopening to trigger the CI
/close

**Review by mstemm — CHANGES_REQUESTED** [review]
@FedeDP could you include the note from the standard about the last element of a struct and how it doesn't change the size of the struct? That's an important consideration.

We'll just need to make sure that part of the standard is honored by most compilers. (We'll need to think of windows as well, as windows does often read trace files).

File: driver/ppm_events_public.h [hunk 1]
```diff
@@ -1485,6 +1485,7 @@ struct ppm_evt_hdr {
 	uint32_t len; /* the event len, including the header */
 	uint16_t type; /* the event type */
 	uint32_t nparams; /* the number of parameters of the event */
+	uint8_t payload[]; /* Flexible array member for event payload, ie: both len headers + event data */
```
[the remark this answers is not in the document]
Comment (FedeDP) [reply driver/ppm_events_public.h:1488]: Included the note!
Flexible array member is part of c99; i think we are still using ansi C89 (or C90)?
Do anyone use compilers that old anymore? I mean, can't we jump to C99?

[the remark this answers is not in the document]
Comment (FedeDP) [reply driver/ppm_events_public.h:1488]: I think C99 is fully supported since gcc 4.3: https://gcc.gnu.org/c99status.html

**Comment by FedeDP** [comment]
I think i will split this PR in:
* a PR that adds the flexible array member to struct ppm_evt_hdr to manage events payload + enforces c99 + merges g_syscall_code_routing_table into g_syscall_table
* a PR that uses an m4 script to automatically create syscall tables

The latter is quite opinionated and might not be merged soon; the former instead should have the consensus to be merged.

**Comment by leogr** [comment]
> I think i will split this PR in:
> 
> * a PR that adds the flexible array member to struct ppm_evt_hdr to manage events payload + enforces c99 + merges g_syscall_code_routing_table into g_syscall_table
> * a PR that uses an m4 script to automatically create syscall tables
> 
> The latter is quite opinionated and might not be merged soon; the former instead should have the consensus to be merged.

Hey @FedeDP 

I totally agree with splitting this into several PRs. I'm also ok with C99. Moreover, I'd suggest documenting somewhere the C standard we are going to use.

**Comment by FedeDP** [comment]
I split https://github.com/falcosecurity/libs/pull/298 from this one, with the first point above:
* flexible array member to struct ppm_evt_hdr to manage events payload 
* enforces c99 
* merges g_syscall_code_routing_table into g_syscall_table

This PR (the #86) will instead become only the m4 script one.

**Comment by FedeDP** [comment]
Most of this PR was split in multiple ones.
We decided to avoid the m4 script, therefore there is not much else we can do.
The only open points are:
* c99
* flexible array member
that are also part of #298 (but i am not sure will ever be implemented).

**Pull request closed without merging**